LibGit2Sharp.NameConflictException: Failed to write reference 'refs/heads/feature/Morpheus_v12.2': a reference with that name already exists.

Sorry, I'm just starting to look at this but I have an exception that occurs only on 1 branch that we have.  The rest are working.  

This is in version 3.5.3 building on a TFS vNext agent.

`ERROR [05/18/16 11:44:42:43] An unexpected error occurred:

```
2016-05-17T23:44:42.4352442Z LibGit2Sharp.NameConflictException: Failed to write reference 'refs/heads/feature/Morpheus_v12.2': a reference with that name already exists.
 2016-05-17T23:44:42.4352442Z    at LibGit2Sharp.Core.Ensure.HandleError(Int32 result)
 2016-05-17T23:44:42.4352442Z    at LibGit2Sharp.Core.Ensure.ZeroResult(Int32 result)
 2016-05-17T23:44:42.4352442Z    at LibGit2Sharp.Core.Proxy.git_reference_create(RepositoryHandle repo, String name, ObjectId targetId, Boolean allowOverwrite, String logMessage)
2016-05-17T23:44:42.4352442Z    at LibGit2Sharp.ReferenceCollection.Add(String name, ObjectId targetId, String logMessage, Boolean allowOverwrite)
2016-05-17T23:44:42.4352442Z    at LibGit2Sharp.ReferenceCollection.Add(String name, ObjectId targetId)
2016-05-17T23:44:42.4352442Z    at GitTools.Git.GitRepositoryHelper.EnsureLocalBranchExistsForCurrentBranch(Repository repo, String currentBranch)
2016-05-17T23:44:42.4352442Z    at GitTools.Git.GitRepositoryHelper.NormalizeGitDirectory(String gitDirectory, AuthenticationInfo authentication, Boolean noFetch, String currentBranch)
2016-05-17T23:44:42.4352442Z    at GitVersion.GitPreparer.Initialise(Boolean normaliseGitDirectory, String currentBranch)
2016-05-17T23:44:42.4352442Z    at GitVersion.ExecuteCore.ExecuteInternal(String targetBranch, String commitId, IRepository repo, GitPreparer gitPreparer, String projectRoot, IBuildServer buildServer, Config overrideConfig)
2016-05-17T23:44:42.4352442Z    at GitVersion.ExecuteCore.ExecuteGitVersion(String targetUrl, String dynamicRepositoryLocation, Authentication authentication, String targetBranch, Boolean noFetch, String workingDirectory, String commitId, Config overrideConfig)
2016-05-17T23:44:42.4352442Z    at GitVersion.SpecifiedArgumentRunner.Run(Arguments arguments, IFileSystem fileSystem)
2016-05-17T23:44:42.4352442Z    at GitVersion.Program.VerifyArgumentsAndRun()
2016-05-17T23:44:42.4352442Z INFO [05/18/16 11:44:42:43] 
2016-05-17T23:44:42.4352442Z INFO [05/18/16 11:44:42:43] Here is the current git graph (please include in issue): 
2016-05-17T23:44:42.4352442Z INFO [05/18/16 11:44:42:43] Showing max of 100 commits
2016-05-17T23:44:43.3258635Z INFO [05/18/16 11:44:43:32] *   e1e8217 47 minutes ago  (HEAD, origin/Feature/Morpheus_v12.2)
2016-05-17T23:44:43.3258635Z |\  
2016-05-17T23:44:43.3258635Z | | * 497957a 58 minutes ago  (origin/pull/US751)
2016-05-17T23:44:43.3414884Z | | *   daf77dc 2 hours ago 
2016-05-17T23:44:43.3414884Z | | |\  
2016-05-17T23:44:43.3414884Z | |_|/  
2016-05-17T23:44:43.3414884Z |/| |   
2016-05-17T23:44:43.3414884Z * | |   4d340a3 2 hours ago 
2016-05-17T23:44:43.3414884Z |\ \ \  
2016-05-17T23:44:43.3414884Z | * \ \   f5d110e 21 hours ago  (Feature/Morpheus_v12.2)
2016-05-17T23:44:43.3414884Z | |\ \ \  
2016-05-17T23:44:43.3414884Z | | * \ \   7c614e2 2 days ago  (origin/pull/US968, pull/US968)
2016-05-17T23:44:43.3414884Z | | |\ \ \  
2016-05-17T23:44:43.3414884Z | | * | | | 4b6cb49 2 days ago 
2016-05-17T23:44:43.3414884Z * | | | | | 8858504 2 hours ago 
2016-05-17T23:44:43.3414884Z |/ / / / /  
2016-05-17T23:44:43.3414884Z | | | | * 853b994 2 hours ago 
2016-05-17T23:44:43.3414884Z | | | | * 5245501 21 hours ago 
2016-05-17T23:44:43.3414884Z | |_|_|/  
2016-05-17T23:44:43.3414884Z |/| | |   
2016-05-17T23:44:43.3414884Z * | | |   0ca06fd 22 hours ago 
2016-05-17T23:44:43.3414884Z |\ \ \ \  
2016-05-17T23:44:43.3414884Z | |_|/ /  
2016-05-17T23:44:43.3414884Z |/| | |   
2016-05-17T23:44:43.3414884Z | | | | *   32923d9 3 hours ago  (origin/Feature/VRecalls)
2016-05-17T23:44:43.3414884Z | | | | |\  
2016-05-17T23:44:43.3414884Z | | | | |/  
2016-05-17T23:44:43.3571133Z | | | |/|   
2016-05-17T23:44:43.3571133Z | | | * |   7aa63d9 3 hours ago  (origin/master, master)
2016-05-17T23:44:43.3571133Z | | | |\ \  
2016-05-17T23:44:43.3571133Z | | | | * | 027a9e5 3 hours ago 
2016-05-17T23:44:43.3571133Z | | | |/ /  
2016-05-17T23:44:43.3571133Z | | | * |   9625f39 4 hours ago 
2016-05-17T23:44:43.3571133Z | | | |\ \  
2016-05-17T23:44:43.3571133Z | | | | * | 2da01c4 4 hours ago 
2016-05-17T23:44:43.3571133Z | | | * | |   3e808dc 19 hours ago 
2016-05-17T23:44:43.3571133Z | | | |\ \ \  
2016-05-17T23:44:43.3571133Z | | | | * | | a87dd32 20 hours ago 
2016-05-17T23:44:43.3571133Z | | | | * | |   279f6cf 20 hours ago 
2016-05-17T23:44:43.3571133Z | | | | |\ \ \  
2016-05-17T23:44:43.3571133Z | | | | | * | | 1eb4bd6 20 hours ago 
2016-05-17T23:44:43.3571133Z | | |_|_|/ / /  
2016-05-17T23:44:43.3571133Z | |/| | | | |   
2016-05-17T23:44:43.3571133Z | | | | * | | 4b846e7 20 hours ago 
2016-05-17T23:44:43.3571133Z | | |_|/ / /  
2016-05-17T23:44:43.3571133Z | |/| | | |   
2016-05-17T23:44:43.3571133Z | | | * | |   36918e5 20 hours ago 
2016-05-17T23:44:43.3571133Z | | | |\ \ \  
2016-05-17T23:44:43.3571133Z | | |/ / / /  
2016-05-17T23:44:43.3571133Z | |/| | | |   
2016-05-17T23:44:43.3571133Z | | | * | | 01e3c52 21 hours ago 
2016-05-17T23:44:43.3571133Z | | |/ / /  
2016-05-17T23:44:43.3571133Z | |/| | |   
2016-05-17T23:44:43.3571133Z | | | | *   9cc356b 21 hours ago  (Feature/VRecalls)
2016-05-17T23:44:43.3571133Z | | | | |\  
2016-05-17T23:44:43.3571133Z | | | | | * de88a2a 21 hours ago 
2016-05-17T23:44:43.3571133Z | | | | | *   4ad26d2 21 hours ago 
2016-05-17T23:44:43.3571133Z | | | | | |\  
2016-05-17T23:44:43.3571133Z | | | | | | *   dfb8e2a 22 hours ago 
2016-05-17T23:44:43.3571133Z | | | | | | |\  
2016-05-17T23:44:43.3571133Z | | |_|_|_|_|/  
2016-05-17T23:44:43.3571133Z | |/| | | | |   
2016-05-17T23:44:43.3571133Z | * | | | | |   726e6cb 24 hours ago 
2016-05-17T23:44:43.3571133Z | |\ \ \ \ \ \  
2016-05-17T23:44:43.3571133Z | | * | | | | | 5856f76 25 hours ago  (origin/pull/M49638, pull/M49638)
2016-05-17T23:44:43.3727382Z | | * | | | | | c9f186c 2 days ago 
2016-05-17T23:44:43.3727382Z | * | | | | | |   53a095b 24 hours ago 
2016-05-17T23:44:43.3727382Z | |\ \ \ \ \ \ \  
2016-05-17T23:44:43.3727382Z | | * | | | | | | a71dff4 2 days ago  (origin/pull/lightinstaller, pull/lightinstaller)
2016-05-17T23:44:43.3727382Z | | * | | | | | | 14b9c79 2 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | | 11e2e91 2 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | | 78c956a 2 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | | cae8e9d 2 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | | 7f08e37 2 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | | 3c3d76a 2 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | | a4ffa79 2 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | | d6d5d4b 2 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | | be6aef8 2 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | | 3c16f81 5 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | | c4a47f8 5 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | | 0d1e995 5 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | | bbcc78c 5 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | | 53680e0 5 days ago 
2016-05-17T23:44:43.3727382Z | | * | | | | | |   e5cdff6 6 days ago 
2016-05-17T23:44:43.3727382Z | | |\ \ \ \ \ \ \  
2016-05-17T23:44:43.3727382Z | | | | |/ / / / /  
2016-05-17T23:44:43.3727382Z | | | |/| | | | |   
2016-05-17T23:44:43.3727382Z | | * | | | | | | 47c468d 6 days ago 
2016-05-17T23:44:43.3727382Z | * | | | | | | |   e5850f1 25 hours ago 
2016-05-17T23:44:43.3727382Z | |\ \ \ \ \ \ \ \  
2016-05-17T23:44:43.3727382Z | | |_|_|_|/ / / /  
2016-05-17T23:44:43.3727382Z | |/| | | | | | |   
2016-05-17T23:44:43.3727382Z | * | | | | | | |   c2eb9ac 26 hours ago 
2016-05-17T23:44:43.3727382Z | |\ \ \ \ \ \ \ \  
2016-05-17T23:44:43.3727382Z | | * \ \ \ \ \ \ \   628b35b 2 days ago  (origin/pull/M49556, pull/M49556)
2016-05-17T23:44:43.3727382Z | | |\ \ \ \ \ \ \ \  
2016-05-17T23:44:43.3727382Z | | |/ / / / / / / /  
2016-05-17T23:44:43.3727382Z | |/| | | | | | | |   
2016-05-17T23:44:43.3727382Z | | * | | | | | | | df15870 6 weeks ago  (origin/M49556, M49556)
2016-05-17T23:44:43.3727382Z | | | | | | | | * |   892a7c0 23 hours ago 
2016-05-17T23:44:43.3727382Z | | | | | | | | |\ \  
2016-05-17T23:44:43.3727382Z | | | | | | | | | |/  
2016-05-17T23:44:43.3727382Z | | | | | | | | * |   a119604 25 hours ago 
2016-05-17T23:44:43.3883631Z | | | | | | | | |\ \  
2016-05-17T23:44:43.3883631Z | | | | | | | | * | | 392e639 25 hours ago 
2016-05-17T23:44:43.3883631Z | | | | | | | * | | | 602ab82 21 hours ago 
2016-05-17T23:44:43.3883631Z | | | | | | | | |_|/  
2016-05-17T23:44:43.3883631Z | | | | | | | |/| |   
2016-05-17T23:44:43.3883631Z | | | | | | | * | |   d31cda9 24 hours ago 
2016-05-17T23:44:43.3883631Z | | | | | | | |\ \ \  
2016-05-17T23:44:43.3883631Z | | | | | | | | * \ \   6565e63 24 hours ago 
2016-05-17T23:44:43.3883631Z | | | | | | | | |\ \ \  
2016-05-17T23:44:43.3883631Z | | | | | | | | | * \ \   b6e60fc 2 days ago 
2016-05-17T23:44:43.3883631Z | | | | | | | | | |\ \ \  
2016-05-17T23:44:43.3883631Z | | |_|_|_|_|_|_|_|/ / /  
2016-05-17T23:44:43.3883631Z | |/| | | | | | | | | |   
2016-05-17T23:44:43.3883631Z | * | | | | | | | | | |   04518f5 2 days ago  (origin/Feature/MedicalHistoryFix, origin/Feature/M47228-DAFAlerts, Feature/MedicalHistoryFix, Feature/M47228-DAFAlerts)
2016-05-17T23:44:43.3883631Z | |\ \ \ \ \ \ \ \ \ \ \  
2016-05-17T23:44:43.3883631Z | * \ \ \ \ \ \ \ \ \ \ \   24a1fc5 2 days ago 
2016-05-17T23:44:43.3883631Z | |\ \ \ \ \ \ \ \ \ \ \ \  
2016-05-17T23:44:43.3883631Z | | |_|_|_|_|_|/ / / / / /  
2016-05-17T23:44:43.3883631Z | |/| | | | | | | | | | |   
2016-05-17T23:44:43.3883631Z | | | | | | | | | | * | | 1d0ce6c 2 days ago 
2016-05-17T23:44:43.3883631Z | | |_|_|_|_|_|_|_|/ / /  
2016-05-17T23:44:43.3883631Z | |/| | | | | | | | | |   
2016-05-17T23:44:43.3883631Z | | | | | | | | * | | | 9cdedf7 24 hours ago 
2016-05-17T23:44:43.3883631Z | | | | | | | | |/ / /  
2016-05-17T23:44:43.3883631Z | | | | | | | | * | | 9a185e6 24 hours ago 
2016-05-17T23:44:43.3883631Z | | | | | | | | | |/  
2016-05-17T23:44:43.3883631Z | | | | | | | | |/|   
2016-05-17T23:44:43.3883631Z | | | | | | | | * | 874d120 25 hours ago 
2016-05-17T23:44:43.3883631Z | | | | | | | | * | 4d72ce0 2 days ago 
2016-05-17T23:44:43.3883631Z | | | | | | | | |/  
2016-05-17T23:44:43.3883631Z | | | | | | | | * 4f445a3 2 days ago 
2016-05-17T23:44:43.3883631Z | | |_|_|_|_|_|/  
2016-05-17T23:44:43.3883631Z | |/| | | | | |   
2016-05-17T23:44:43.3883631Z | | | | | | | | * 2289d18 20 hours ago  (origin/pull/TFS-949, pull/TFS-949)
2016-05-17T23:44:43.3883631Z | |_|_|_|_|_|_|/  
2016-05-17T23:44:43.3883631Z |/| | | | | | |   
2016-05-17T23:44:43.3883631Z * | | | | | | |   94f3b17 2 days ago 
2016-05-17T23:44:43.4039880Z |\ \ \ \ \ \ \ \  
2016-05-17T23:44:43.4039880Z | * | | | | | | | 2412edc 5 days ago  (origin/pull/US962, pull/US962)
2016-05-17T23:44:43.4039880Z | | |_|_|_|_|_|/  
2016-05-17T23:44:43.4039880Z | |/| | | | | |   
2016-05-17T23:44:43.4039880Z * | | | | | | |   48ae8c6 2 days ago 
2016-05-17T23:44:43.4039880Z |\ \ \ \ \ \ \ \  
2016-05-17T23:44:43.4039880Z | |/ / / / / / /  
2016-05-17T23:44:43.4039880Z |/| | | | | | |   

2016-05-17T23:44:43.4039880Z | * | | | | | | fdc04f5 2 days ago  (origin/pull/US966, pull/US966)
 <snip>
```

I am having this on our master branch in one of our projects. We have 5 other Azure DevOps pipeline builds running fine. But this 6th build is giving an error:
`##[error]C:\Users\VssAdministrator\.nuget\packages\gitversiontask\5.2.4\build\GitVersionTask.targets(10,9): Error : NameConflictException: failed to write reference 'refs/heads/master': a reference with that name already exists.
`
This branch obviously always has been called "master", so it is not a casing issue in our case.
We have PR triggered builds turned off. We trigger builds on every checkin. So in our case it is not on PR builds.

Our VS solution has 3 projects with "GitVersionTask" installed. When I remove the "GitVersionTask" from 1 project (an leave it in the other 2). The build succeeds. So the task works fine in the other 2 projects. If I then add the "GitVersionTask" back to the 3th project, the build still succeeds. But when I then make any other change/checkin, the build fails again with the same error message. 

Does this make any sense to anyone?

--- a/src/GitVersionCore.Tests/Core/RepositoryExtensionsTests.cs
+++ b/src/GitVersionCore.Tests/Core/RepositoryExtensionsTests.cs
@@ -1,0 +1,188 @@
+using GitTools.Testing;
+using GitVersion.Logging;
+using GitVersion.Extensions;
+using GitVersionCore.Tests.Helpers;
+using LibGit2Sharp;
+using NUnit.Framework;
+using System.Linq;
+using NSubstitute;
+using System;
+using System.Collections.Generic;
+
+namespace GitVersionCore.Tests
+{
+    [TestFixture]
+    public class RepositoryExtensionsTests : TestBase
+    {
+        [Test]
+        public void EnsureLocalBranchExistsForCurrentBranch_CaseInsensitivelyMatchesBranches()
+        {
+            ILog log = Substitute.For<ILog>();
+            var repository = Substitute.For<IRepository>();
+            var remote = ConstructRemote(repository);
+
+            repository.EnsureLocalBranchExistsForCurrentBranch(log, remote, "refs/heads/featurE/feat-test");
+        }
+
+        private Remote ConstructRemote(IRepository repository)
+        {
+            var branches = new TestableBranchCollection(repository);
+            var tipId = new ObjectId("c6d8764d20ff16c0df14c73680e52b255b608926");
+            var tip = new TestableCommit(repository, tipId);
+            var head = branches.Add("refs/heads/feature/feat-test", tip);
+            var remote = new TesatbleRemote("origin");
+            var references = new TestableReferenceCollection();
+            var reference = references.Add("develop", "refs/heads/develop");
+
+            repository.Refs.Returns(references);
+            repository.Head.Returns(head);
+            repository.Branches.Returns(branches);
+            return remote;
+        }
+
+        private class TestableBranchCollection : BranchCollection
+        {
+            private readonly IRepository repository;
+            public TestableBranchCollection(IRepository repository)
+            {
+            }
+
+            IDictionary<string, Branch> branches = new Dictionary<string, Branch>();
+
+            public override Branch this[string name] =>
+                this.branches.ContainsKey(name)
+                    ? this.branches[name]
+                    : null;
+
+            public override Branch Add(string name, Commit commit)
+            {
+                var branch = new TestableBranch(name, commit);
+                this.branches.Add(name, branch);
+                return branch;
+            }
+
+            public override Branch Add(string name, string committish)
+            {
+                var id = new ObjectId(committish);
+                var commit = new TestableCommit(this.repository, id);
+                return Add(name, commit);
+            }
+
+            public override Branch Add(string name, Commit commit, bool allowOverwrite)
+            {
+                return Add(name, commit);
+            }
+
+            public override Branch Add(string name, string committish, bool allowOverwrite)
+            {
+                return Add(name, committish);
+            }
+
+            public override IEnumerator<Branch> GetEnumerator()
+            {
+                return this.branches.Values.GetEnumerator();
+            }
+
+            public override void Remove(string name)
+            {
+                this.branches.Remove(name);
+            }
+
+            public override void Remove(string name, bool isRemote)
+            {
+                this.branches.Remove(name);
+            }
+
+            public override void Remove(Branch branch)
+            {
+                this.branches.Remove(branch.CanonicalName);
+            }
+
+            public override Branch Update(Branch branch, params Action<BranchUpdater>[] actions)
+            {
+                return base.Update(branch, actions);
+            }
+        }
+
+        private class TestableBranch : Branch
+        {
+            private readonly string canonicalName;
+            private readonly Commit tip;
+
+            public TestableBranch(string canonicalName, Commit tip)
+            {
+                this.tip = tip;
+                this.canonicalName = canonicalName;
+            }
+
+            public override string CanonicalName => this.canonicalName;
+            public override Commit Tip => this.tip;
+        }
+
+        private class TestableCommit : Commit, IBelongToARepository
+        {
+            private IRepository repository;
+            private ObjectId id;
+
+            public TestableCommit(IRepository repository, ObjectId id)
+            {
+                this.repository = repository;
+                this.id = id;
+            }
+
+            public override ObjectId Id => this.id;
+            public IRepository Repository => this.repository;
+        }
+
+        private class TesatbleRemote : Remote
+        {
+            private string name;
+
+            public TesatbleRemote(string name)
+            {
+                this.name = name;
+            }
+
+            public override string Name => this.name;
+        }
+
+        private class TestableReferenceCollection : ReferenceCollection
+        {
+            Reference reference;
+
+            public override DirectReference Add(string name, ObjectId targetId)
+            {
+                throw new InvalidOperationException("Update should be invoked when case-insensitively comparing branches.");
+            }
+
+            public override Reference Add(string name, string canonicalRefNameOrObjectish)
+            {
+                return this.reference = new TestableReference(canonicalRefNameOrObjectish);;
+            }
+
+            public override Reference UpdateTarget(Reference directRef, ObjectId targetId)
+            {
+                return this.reference;
+            }
+
+            public override Reference this[string name] => this.reference;
+        }
+
+        private class TestableReference : Reference
+        {
+            private readonly string canonicalName;
+
+            public TestableReference(string canonicalName)
+            {
+                this.canonicalName = canonicalName;
+            }
+
+            public override string CanonicalName => this.canonicalName;
+
+            public override DirectReference ResolveToDirectReference()
+            {
+                throw new NotImplementedException();
+            }
+        }
+    }
+}

--- a/src/GitVersionCore.Tests/Core/RepositoryExtensionsTests.cs
+++ b/src/GitVersionCore.Tests/Core/RepositoryExtensionsTests.cs
@@ -8,6 +8,7 @@ using System.Linq;
 using NSubstitute;
 using System;
 using System.Collections.Generic;
+using GitVersion;
 
 namespace GitVersionCore.Tests
 {
@@ -17,14 +18,22 @@ namespace GitVersionCore.Tests
         [Test]
         public void EnsureLocalBranchExistsForCurrentBranch_CaseInsensitivelyMatchesBranches()
         {
-            ILog log = Substitute.For<ILog>();
-            var repository = Substitute.For<IRepository>();
-            var remote = ConstructRemote(repository);
+            var log = Substitute.For<ILog>();
+            var repository = MockRepository();
+            var remote = MockRemote(repository);
 
             repository.EnsureLocalBranchExistsForCurrentBranch(log, remote, "refs/heads/featurE/feat-test");
         }
 
-        private Remote ConstructRemote(IRepository repository)
+        private IGitRepository MockRepository()
+        {
+            var repository = Substitute.For<IGitRepository>();
+            var commands = Substitute.For<IGitRepositoryCommands>();
+            repository.Commands.Returns(commands);
+            return repository;
+        }
+
+        private Remote MockRemote(IGitRepository repository)
         {
             var branches = new TestableBranchCollection(repository);
             var tipId = new ObjectId("c6d8764d20ff16c0df14c73680e52b255b608926");
@@ -157,7 +166,7 @@ namespace GitVersionCore.Tests
 
             public override Reference Add(string name, string canonicalRefNameOrObjectish)
             {
-                return this.reference = new TestableReference(canonicalRefNameOrObjectish);;
+                return this.reference = new TestableReference(canonicalRefNameOrObjectish);
             }
 
             public override Reference UpdateTarget(Reference directRef, ObjectId targetId)

--- a/src/GitVersionCore/Configuration/BranchConfigurationCalculator.cs
+++ b/src/GitVersionCore/Configuration/BranchConfigurationCalculator.cs
@@ -41,7 +41,7 @@ namespace GitVersion.Configuration
             if (matchingBranches.Increment == IncrementStrategy.Inherit)
             {
                 matchingBranches = InheritBranchConfiguration(targetBranch, matchingBranches, currentCommit, configuration, excludedInheritBranches);
-                if (matchingBranches.Name == FallbackConfigName && matchingBranches.Increment == IncrementStrategy.Inherit)
+                if (matchingBranches.Name.IsEquivalentTo(FallbackConfigName) && matchingBranches.Increment == IncrementStrategy.Inherit)
                 {
                     // We tried, and failed to inherit, just fall back to patch
                     matchingBranches.Increment = IncrementStrategy.Patch;
@@ -152,7 +152,7 @@ namespace GitVersion.Configuration
 
                 var inheritingBranchConfig = GetBranchConfiguration(chosenBranch, currentCommit, configuration, excludedInheritBranches);
                 var configIncrement = inheritingBranchConfig.Increment;
-                if (inheritingBranchConfig.Name == FallbackConfigName && configIncrement == IncrementStrategy.Inherit)
+                if (inheritingBranchConfig.Name.IsEquivalentTo(FallbackConfigName) && configIncrement == IncrementStrategy.Inherit)
                 {
                     log.Warning("Fallback config inherits by default, dropping to patch increment");
                     configIncrement = IncrementStrategy.Patch;

--- a/src/GitVersionCore/Core/GitPreparer.cs
+++ b/src/GitVersionCore/Core/GitPreparer.cs
@@ -202,7 +202,7 @@ namespace GitVersion
         private void NormalizeGitDirectory(string gitDirectory, bool noFetch, string currentBranch, bool isDynamicRepository)
         {
             var authentication = options.Value.Authentication;
-            using var repository = new Repository(gitDirectory);
+            using var repository = new GitRepository(() => gitDirectory);
             // Need to ensure the HEAD does not move, this is essentially a BugCheck
             var expectedSha = repository.Head.Tip.Sha;
             var expectedBranchName = repository.Head.CanonicalName;
@@ -221,7 +221,7 @@ namespace GitVersion
                 else
                 {
                     log.Info($"Fetching from remote '{remote.Name}' using the following refspecs: {string.Join(", ", remote.FetchRefSpecs.Select(r => r.Specification))}.");
-                    Commands.Fetch(repository, remote.Name, new string[0], authentication.ToFetchOptions(), null);
+                    repository.Commands.Fetch(remote.Name, new string[0], authentication.ToFetchOptions(), null);
                 }
 
                 repository.EnsureLocalBranchExistsForCurrentBranch(log, remote, currentBranch);
@@ -262,7 +262,7 @@ namespace GitVersion
                 if (matchingCurrentBranch != null)
                 {
                     log.Info($"Checking out local branch '{currentBranch}'.");
-                    Commands.Checkout(repository, matchingCurrentBranch);
+                    repository.Commands.Checkout(matchingCurrentBranch);
                 }
                 else if (localBranchesWhereCommitShaIsHead.Count > 1)
                 {
@@ -275,7 +275,7 @@ namespace GitVersion
                     if (master != null)
                     {
                         log.Warning("Because one of the branches is 'master', will build master." + moveBranchMsg);
-                        Commands.Checkout(repository, master);
+                        repository.Commands.Checkout(master);
                     }
                     else
                     {
@@ -284,7 +284,7 @@ namespace GitVersion
                         {
                             var branchWithoutSeparator = branchesWithoutSeparators[0];
                             log.Warning($"Choosing {branchWithoutSeparator.CanonicalName} as it is the only branch without / or - in it. " + moveBranchMsg);
-                            Commands.Checkout(repository, branchWithoutSeparator);
+                            repository.Commands.Checkout(branchWithoutSeparator);
                         }
                         else
                         {
@@ -300,7 +300,7 @@ namespace GitVersion
                 else
                 {
                     log.Info($"Checking out local branch 'refs/heads/{localBranchesWhereCommitShaIsHead[0].FriendlyName}'.");
-                    Commands.Checkout(repository, repository.Branches[localBranchesWhereCommitShaIsHead[0].FriendlyName]);
+                    repository.Commands.Checkout(repository.Branches[localBranchesWhereCommitShaIsHead[0].FriendlyName]);
                 }
             }
             finally

--- a/src/GitVersionCore/Core/GitPreparer.cs
+++ b/src/GitVersionCore/Core/GitPreparer.cs
@@ -271,7 +271,7 @@ namespace GitVersion
                     const string moveBranchMsg = "Move one of the branches along a commit to remove warning";
 
                     log.Warning($"Found more than one local branch pointing at the commit '{headSha}' ({csvNames}).");
-                    var master = localBranchesWhereCommitShaIsHead.SingleOrDefault(n => n.FriendlyName == "master");
+                    var master = localBranchesWhereCommitShaIsHead.SingleOrDefault(n => n.FriendlyName.IsEquivalentTo("master"));
                     if (master != null)
                     {
                         log.Warning("Because one of the branches is 'master', will build master." + moveBranchMsg);

--- a/src/GitVersionCore/Core/GitRepository.cs
+++ b/src/GitVersionCore/Core/GitRepository.cs
@@ -5,14 +5,20 @@ using Microsoft.Extensions.Options;
 
 namespace GitVersion
 {
-    public class GitRepository : IRepository
+    public class GitRepository : IGitRepository
     {
         private Lazy<IRepository> repositoryLazy;
         private IRepository repositoryInstance => repositoryLazy.Value;
 
         public GitRepository(IOptions<GitVersionOptions> options)
+            : this(() => options.Value.DotGitDirectory)
         {
-            repositoryLazy = new Lazy<IRepository>(() => new Repository(options.Value.DotGitDirectory));
+        }
+
+        public GitRepository(Func<string> getDotGitDirectory)
+        {
+            repositoryLazy = new Lazy<IRepository>(() => new Repository(getDotGitDirectory()));
+            Commands = new GitRepositoryCommands(repositoryLazy);
         }
 
         public void Dispose()
@@ -161,5 +167,7 @@ namespace GitVersion
         public Network Network => repositoryInstance.Network;
 
         public StashCollection Stashes => repositoryInstance.Stashes;
+
+        public IGitRepositoryCommands Commands { get; }
     }
 }

--- a/src/GitVersionCore/Core/GitRepositoryCommands.cs
+++ b/src/GitVersionCore/Core/GitRepositoryCommands.cs
@@ -1,0 +1,141 @@
+using System;
+using System.Collections.Generic;
+using LibGit2Sharp;
+
+namespace GitVersion
+{
+    /// <summary>
+    /// Default implementation of <see cref="IGitRepositoryCommands"/> using
+    /// the <c>static</c> <see cref="Commands"/> <c>class</c>.
+    /// </summary>
+    public class GitRepositoryCommands : IGitRepositoryCommands
+    {
+        private readonly Lazy<IRepository> lazyRepository;
+        private IRepository repository => lazyRepository.Value;
+
+        public GitRepositoryCommands(Lazy<IRepository> lazyRepository)
+        {
+            this.lazyRepository = lazyRepository ?? throw new System.ArgumentNullException(nameof(lazyRepository));
+        }
+
+        public Branch Checkout(string committishOrBranchSpec)
+        {
+            return Commands.Checkout(this.repository, committishOrBranchSpec);
+        }
+
+        public Branch Checkout(string committishOrBranchSpec, CheckoutOptions options)
+        {
+            return Commands.Checkout(this.repository, committishOrBranchSpec, options);
+        }
+
+        public Branch Checkout(Branch branch)
+        {
+            return Commands.Checkout(this.repository, branch);
+        }
+
+        public Branch Checkout(Branch branch, CheckoutOptions options)
+        {
+            return Commands.Checkout(this.repository, branch, options);
+        }
+
+        public Branch Checkout(Commit commit)
+        {
+            return Commands.Checkout(this.repository, commit);
+        }
+
+        public Branch Checkout(Commit commit, CheckoutOptions options)
+        {
+            return Commands.Checkout(this.repository, commit, options);
+        }
+
+        public void Checkout(Tree tree, CheckoutOptions checkoutOptions, string refLogHeadSpec)
+        {
+            Commands.Checkout(this.repository, tree, checkoutOptions, refLogHeadSpec);
+        }
+
+        public void Fetch(string remote, IEnumerable<string> refspecs, FetchOptions options, string logMessage)
+        {
+            Commands.Fetch((Repository)this.repository, remote, refspecs, options, logMessage);
+        }
+
+        public void Move(string sourcePath, string destinationPath)
+        {
+            Commands.Move(this.repository, sourcePath, destinationPath);
+        }
+
+        public void Move(IEnumerable<string> sourcePaths, IEnumerable<string> destinationPaths)
+        {
+            Commands.Move(this.repository, sourcePaths, destinationPaths);
+        }
+
+        public MergeResult Pull(Signature merger, PullOptions options)
+        {
+            return Commands.Pull((Repository)this.repository, merger, options);
+        }
+
+        public void Remove(string path, bool removeFromWorkingDirectory)
+        {
+            Commands.Remove(this.repository, path, removeFromWorkingDirectory);
+        }
+
+        public void Remove(IEnumerable<string> paths)
+        {
+            Commands.Remove(this.repository, paths);
+        }
+
+        public void Remove(IEnumerable<string> paths, bool removeFromWorkingDirectory, ExplicitPathsOptions explicitPathsOptions)
+        {
+            Commands.Remove(this.repository, paths, removeFromWorkingDirectory, explicitPathsOptions);
+        }
+
+        public void Remove(string path)
+        {
+            Commands.Remove(this.repository, path);
+        }
+
+        public void Remove(string path, bool removeFromWorkingDirectory, ExplicitPathsOptions explicitPathsOptions)
+        {
+            Commands.Remove(this.repository, path, removeFromWorkingDirectory, explicitPathsOptions);
+        }
+
+        public void Stage(string path)
+        {
+            Commands.Stage(this.repository, path);
+        }
+
+        public void Stage(string path, StageOptions stageOptions)
+        {
+            Commands.Stage(this.repository, path, stageOptions);
+        }
+
+        public void Stage(IEnumerable<string> paths)
+        {
+            Commands.Stage(this.repository, paths);
+        }
+
+        public void Stage(IEnumerable<string> paths, StageOptions stageOptions)
+        {
+            Commands.Stage(this.repository, paths, stageOptions);
+        }
+
+        public void Unstage(string path)
+        {
+            Commands.Unstage(this.repository, path);
+        }
+
+        public void Unstage(string path, ExplicitPathsOptions explicitPathsOptions)
+        {
+            Commands.Unstage(this.repository, path, explicitPathsOptions);
+        }
+
+        public void Unstage(IEnumerable<string> paths)
+        {
+            Commands.Unstage(this.repository, paths);
+        }
+
+        public void Unstage(IEnumerable<string> paths, ExplicitPathsOptions explicitPathsOptions)
+        {
+            Commands.Unstage(this.repository, paths, explicitPathsOptions);
+        }
+    }
+}

--- a/src/GitVersionCore/Core/IGitRepository.cs
+++ b/src/GitVersionCore/Core/IGitRepository.cs
@@ -1,0 +1,9 @@
+using LibGit2Sharp;
+
+namespace GitVersion
+{
+    public interface IGitRepository : IRepository
+    {
+        IGitRepositoryCommands Commands { get; }
+    }
+}

--- a/src/GitVersionCore/Core/IGitRepositoryCommands.cs
+++ b/src/GitVersionCore/Core/IGitRepositoryCommands.cs
@@ -1,0 +1,37 @@
+using System.Collections.Generic;
+using LibGit2Sharp;
+
+namespace GitVersion
+{
+    /// <summary>
+    /// Mockable and testable interface wrapper for the <c>static</c>
+    /// <see cref="Commands"/> <c>class</c>.
+    /// </summary>
+    public interface IGitRepositoryCommands
+    {
+        Branch Checkout(string committishOrBranchSpec);
+        Branch Checkout(string committishOrBranchSpec, CheckoutOptions options);
+        Branch Checkout(Branch branch);
+        Branch Checkout(Branch branch, CheckoutOptions options);
+        Branch Checkout(Commit commit);
+        Branch Checkout(Commit commit, CheckoutOptions options);
+        void Checkout(Tree tree, CheckoutOptions checkoutOptions, string refLogHeadSpec);
+        void Fetch(string remote, IEnumerable<string> refspecs, FetchOptions options, string logMessage);
+        void Move(string sourcePath, string destinationPath);
+        void Move(IEnumerable<string> sourcePaths, IEnumerable<string> destinationPaths);
+        MergeResult Pull(Signature merger, PullOptions options);
+        void Remove(string path, bool removeFromWorkingDirectory);
+        void Remove(IEnumerable<string> paths);
+        void Remove(IEnumerable<string> paths, bool removeFromWorkingDirectory, ExplicitPathsOptions explicitPathsOptions);
+        void Remove(string path);
+        void Remove(string path, bool removeFromWorkingDirectory, ExplicitPathsOptions explicitPathsOptions);
+        void Stage(string path);
+        void Stage(string path, StageOptions stageOptions);
+        void Stage(IEnumerable<string> paths);
+        void Stage(IEnumerable<string> paths, StageOptions stageOptions);
+        void Unstage(string path);
+        void Unstage(string path, ExplicitPathsOptions explicitPathsOptions);
+        void Unstage(IEnumerable<string> paths);
+        void Unstage(IEnumerable<string> paths, ExplicitPathsOptions explicitPathsOptions);
+    }
+}

--- a/src/GitVersionCore/Core/RepositoryMetadataProvider.cs
+++ b/src/GitVersionCore/Core/RepositoryMetadataProvider.cs
@@ -184,9 +184,9 @@ namespace GitVersion
                     // In the case where HEAD is not the desired branch, try to find the branch with matching name
                     desiredBranch = repository.Branches?
                         .SingleOrDefault(b =>
-                            b.CanonicalName == targetBranch ||
-                            b.FriendlyName == targetBranch ||
-                            b.NameWithoutRemote() == targetBranch);
+                            b.CanonicalName.IsEquivalentTo(targetBranch) ||
+                            b.FriendlyName.IsEquivalentTo(targetBranch) ||
+                            b.NameWithoutRemote().IsEquivalentTo(targetBranch));
 
                     // Failsafe in case the specified branch is invalid
                     desiredBranch ??= repository.Head;

--- a/src/GitVersionCore/Extensions/LibGitExtensions.cs
+++ b/src/GitVersionCore/Extensions/LibGitExtensions.cs
@@ -38,7 +38,7 @@ namespace GitVersion.Extensions
             var otherBranchFriendlyName = otherBranch.NameWithoutRemote();
             var branchFriendlyName = branch.NameWithoutRemote();
 
-            return otherBranchFriendlyName == branchFriendlyName;
+            return otherBranchFriendlyName.IsEquivalentTo(branchFriendlyName);
         }
 
         /// <summary>

--- a/src/GitVersionCore/Extensions/RepositoryExtensions.cs
+++ b/src/GitVersionCore/Extensions/RepositoryExtensions.cs
@@ -36,11 +36,25 @@ namespace GitVersion.Extensions
 
         public static void EnsureLocalBranchExistsForCurrentBranch(this IRepository repo, ILog log, Remote remote, string currentBranch)
         {
+            if (log is null)
+            {
+                throw new ArgumentNullException(nameof(log));
+            }
+
+            if (remote is null)
+            {
+                throw new ArgumentNullException(nameof(remote));
+            }
+
             if (string.IsNullOrEmpty(currentBranch)) return;
 
             var isRef = currentBranch.Contains("refs");
             var isBranch = currentBranch.Contains("refs/heads");
-            var localCanonicalName = !isRef ? "refs/heads/" + currentBranch : isBranch ? currentBranch : currentBranch.Replace("refs/", "refs/heads/");
+            var localCanonicalName = !isRef
+                ? "refs/heads/" + currentBranch
+                : isBranch
+                    ? currentBranch
+                    : currentBranch.Replace("refs/", "refs/heads/");
 
             var repoTip = repo.Head.Tip;
 

--- a/src/GitVersionCore/Extensions/RepositoryExtensions.cs
+++ b/src/GitVersionCore/Extensions/RepositoryExtensions.cs
@@ -34,7 +34,7 @@ namespace GitVersion.Extensions
             LibGitExtensions.DumpGraph(repository.Info.Path, writer, maxCommits);
         }
 
-        public static void EnsureLocalBranchExistsForCurrentBranch(this IRepository repo, ILog log, Remote remote, string currentBranch)
+        public static void EnsureLocalBranchExistsForCurrentBranch(this IGitRepository repo, ILog log, Remote remote, string currentBranch)
         {
             if (log is null)
             {
@@ -78,10 +78,11 @@ namespace GitVersion.Extensions
             {
                 log.Info(isBranch ? $"Updating local branch {localCanonicalName} to point at {repoTip.Sha}"
                     : $"Updating local branch {localCanonicalName} to match ref {currentBranch}");
-                repo.Refs.UpdateTarget(repo.Refs[localCanonicalName], repoTipId);
+                var localRef = repo.Refs[localCanonicalName];
+                repo.Refs.UpdateTarget(localRef, repoTipId);
             }
 
-            Commands.Checkout(repo, localCanonicalName);
+            repo.Commands.Checkout(localCanonicalName);
         }
 
         public static void AddMissingRefSpecs(this IRepository repo, ILog log, Remote remote)
@@ -97,7 +98,7 @@ namespace GitVersion.Extensions
                 r => r.FetchRefSpecs.Add(allBranchesFetchRefSpec));
         }
 
-        public static void CreateFakeBranchPointingAtThePullRequestTip(this IRepository repo, ILog log, AuthenticationInfo authentication)
+        public static void CreateFakeBranchPointingAtThePullRequestTip(this IGitRepository repo, ILog log, AuthenticationInfo authentication)
         {
             var remote = repo.Network.Remotes.Single();
 
@@ -133,7 +134,7 @@ namespace GitVersion.Extensions
             if (canonicalName.StartsWith("refs/tags"))
             {
                 log.Info($"Checking out tag '{canonicalName}'");
-                Commands.Checkout(repo, reference.Target.Sha);
+                repo.Commands.Checkout(reference.Target.Sha);
                 return;
             }
 
@@ -149,7 +150,7 @@ namespace GitVersion.Extensions
             repo.Refs.Add(fakeBranchName, new ObjectId(headTipSha));
 
             log.Info($"Checking local branch '{fakeBranchName}' out.");
-            Commands.Checkout(repo, fakeBranchName);
+            repo.Commands.Checkout(fakeBranchName);
         }
 
         public static void CreateOrUpdateLocalBranchesFromRemoteTrackingOnes(this IRepository repo, ILog log, string remoteName)

--- a/src/GitVersionCore/Extensions/RepositoryExtensions.cs
+++ b/src/GitVersionCore/Extensions/RepositoryExtensions.cs
@@ -68,7 +68,7 @@ namespace GitVersion.Extensions
 
             var repoTipId = repoTip.Id;
 
-            if (repo.Branches.All(b => b.CanonicalName != localCanonicalName))
+            if (repo.Branches.All(b => !b.CanonicalName.IsEquivalentTo(localCanonicalName)))
             {
                 log.Info(isBranch ? $"Creating local branch {localCanonicalName}"
                     : $"Creating local branch {localCanonicalName} pointing at {repoTipId}");
@@ -156,17 +156,20 @@ namespace GitVersion.Extensions
         {
             var prefix = $"refs/remotes/{remoteName}/";
             var remoteHeadCanonicalName = $"{prefix}HEAD";
+            var remoteTrackingReferences = repo.Refs
+                .FromGlob(prefix + "*")
+                .Where(r => !r.CanonicalName.IsEquivalentTo(remoteHeadCanonicalName));
 
-            foreach (var remoteTrackingReference in repo.Refs.FromGlob(prefix + "*").Where(r => r.CanonicalName != remoteHeadCanonicalName))
+            foreach (var remoteTrackingReference in remoteTrackingReferences)
             {
                 var remoteTrackingReferenceName = remoteTrackingReference.CanonicalName;
                 var branchName = remoteTrackingReferenceName.Substring(prefix.Length);
                 var localCanonicalName = "refs/heads/" + branchName;
 
                 // We do not want to touch our current branch
-                if (branchName == repo.Head.FriendlyName) continue;
+                if (branchName.IsEquivalentTo(repo.Head.FriendlyName)) continue;
 
-                if (repo.Refs.Any(x => x.CanonicalName == localCanonicalName))
+                if (repo.Refs.Any(x => x.CanonicalName.IsEquivalentTo(localCanonicalName)))
                 {
                     var localRef = repo.Refs[localCanonicalName];
                     var remotedirectReference = remoteTrackingReference.ResolveToDirectReference();

--- a/src/GitVersionCore/Extensions/StringExtensions.cs
+++ b/src/GitVersionCore/Extensions/StringExtensions.cs
@@ -129,5 +129,10 @@ namespace GitVersion.Extensions
         {
             return Regex.Replace(input, pattern, replace, options);
         }
+
+        public static bool IsEquivalentTo(this string self, string other)
+        {
+            return String.Equals(self, other, StringComparison.OrdinalIgnoreCase);
+        }
     }
 }

--- a/src/GitVersionCore/VersionCalculation/NextVersionCalculator.cs
+++ b/src/GitVersionCore/VersionCalculation/NextVersionCalculator.cs
@@ -109,7 +109,7 @@ namespace GitVersion.VersionCalculation
 
             var lastTag = repositoryMetadataProvider
                 .GetVersionTagsOnBranch(context.CurrentBranch, context.Configuration.GitTagPrefix)
-                .FirstOrDefault(v => v.PreReleaseTag.Name == tagToUse);
+                .FirstOrDefault(v => v.PreReleaseTag.Name.IsEquivalentTo(tagToUse));
 
             if (lastTag != null &&
                 MajorMinorPatchEqual(lastTag, semanticVersion) &&

